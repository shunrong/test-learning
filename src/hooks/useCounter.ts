/**
 * 计数器 Hook - 用于演示自定义 Hook 的测试
 */

import { useState, useCallback } from "react";

export interface UseCounterOptions {
  initialValue?: number;
  min?: number;
  max?: number;
  step?: number;
}

export interface UseCounterReturn {
  count: number;
  increment: () => void;
  decrement: () => void;
  reset: () => void;
  set: (value: number) => void;
  canIncrement: boolean;
  canDecrement: boolean;
}

export const useCounter = (
  options: UseCounterOptions = {}
): UseCounterReturn => {
  const {
    initialValue = 0,
    min = -Infinity,
    max = Infinity,
    step = 1,
  } = options;

  const [count, setCount] = useState(initialValue);

  const increment = useCallback(() => {
    setCount((prevCount) => {
      const newCount = prevCount + step;
      return newCount <= max ? newCount : prevCount;
    });
  }, [step, max]);

  const decrement = useCallback(() => {
    setCount((prevCount) => {
      const newCount = prevCount - step;
      return newCount >= min ? newCount : prevCount;
    });
  }, [step, min]);

  const reset = useCallback(() => {
    setCount(initialValue);
  }, [initialValue]);

  const set = useCallback(
    (value: number) => {
      if (value >= min && value <= max) {
        setCount(value);
      }
    },
    [min, max]
  );

  const canIncrement = count + step <= max;
  const canDecrement = count - step >= min;

  return {
    count,
    increment,
    decrement,
    reset,
    set,
    canIncrement,
    canDecrement,
  };
};
