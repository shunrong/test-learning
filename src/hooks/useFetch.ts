/**
 * Fetch Hook - 用于演示异步数据获取和错误处理的测试
 */

import { useState, useEffect, useCallback, useRef } from "react";

export interface UseFetchState<T> {
  data: T | null;
  loading: boolean;
  error: Error | null;
}

export interface UseFetchOptions {
  immediate?: boolean;
  timeout?: number;
}

export interface UseFetchReturn<T> extends UseFetchState<T> {
  execute: () => Promise<void>;
  reset: () => void;
}

export function useFetch<T>(
  url: string,
  options: UseFetchOptions = {}
): UseFetchReturn<T> {
  const { immediate = true, timeout = 5000 } = options;

  const [state, setState] = useState<UseFetchState<T>>({
    data: null,
    loading: false,
    error: null,
  });

  const abortControllerRef = useRef<AbortController | null>(null);

  const execute = useCallback(async () => {
    // 取消之前的请求
    if (abortControllerRef.current) {
      abortControllerRef.current.abort();
    }

    abortControllerRef.current = new AbortController();
    const controller = abortControllerRef.current;

    setState((prev) => ({ ...prev, loading: true, error: null }));

    let timeoutId: NodeJS.Timeout | undefined;

    try {
      timeoutId = setTimeout(() => controller.abort(), timeout);

      const response = await fetch(url, {
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const data = await response.json();

      // 检查请求是否被取消
      if (!controller.signal.aborted) {
        setState({ data, loading: false, error: null });
      }
    } catch (error) {
      if (timeoutId) clearTimeout(timeoutId);

      if (error instanceof Error) {
        if (error.name === "AbortError") {
          setState((prev) => ({
            ...prev,
            loading: false,
            error: new Error("Request timed out"),
          }));
        } else {
          setState((prev) => ({ ...prev, loading: false, error }));
        }
      } else {
        setState((prev) => ({
          ...prev,
          loading: false,
          error: new Error("Unknown error occurred"),
        }));
      }
    }
  }, [url, timeout]);

  const reset = useCallback(() => {
    setState({
      data: null,
      loading: false,
      error: null,
    });
  }, []);

  useEffect(() => {
    if (immediate) {
      execute();
    }

    return () => {
      if (abortControllerRef.current) {
        abortControllerRef.current.abort();
      }
    };
  }, [execute, immediate]);

  return {
    ...state,
    execute,
    reset,
  };
}
