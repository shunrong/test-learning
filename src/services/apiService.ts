/**
 * API 服务 - 用于演示异步测试和 Mock
 */

import axios, { AxiosResponse } from "axios";
import { User, ApiResponse } from "@/types";

// 创建 axios 实例
const apiClient = axios.create({
  baseURL: "https://jsonplaceholder.typicode.com",
  timeout: 5000,
  headers: {
    "Content-Type": "application/json",
  },
});

// 请求拦截器
apiClient.interceptors.request.use(
  (config) => {
    // 添加认证 token
    const token = localStorage.getItem("authToken");
    if (token) {
      config.headers.Authorization = `Bearer ${token}`;
    }
    return config;
  },
  (error) => {
    return Promise.reject(error);
  }
);

// 响应拦截器
apiClient.interceptors.response.use(
  (response) => response,
  (error) => {
    if (error.response?.status === 401) {
      // 处理未授权错误
      localStorage.removeItem("authToken");
      window.location.href = "/login";
    }
    return Promise.reject(error);
  }
);

// 用户相关 API
export const userService = {
  // 获取所有用户
  async getUsers(): Promise<User[]> {
    try {
      const response: AxiosResponse<User[]> = await apiClient.get("/users");
      return response.data;
    } catch (error) {
      throw new Error("Failed to fetch users");
    }
  },

  // 根据 ID 获取用户
  async getUserById(id: number): Promise<User> {
    try {
      const response: AxiosResponse<User> = await apiClient.get(`/users/${id}`);
      return response.data;
    } catch (error) {
      throw new Error(`Failed to fetch user with id ${id}`);
    }
  },

  // 创建用户
  async createUser(user: Omit<User, "id">): Promise<User> {
    try {
      const response: AxiosResponse<User> = await apiClient.post(
        "/users",
        user
      );
      return response.data;
    } catch (error) {
      throw new Error("Failed to create user");
    }
  },

  // 更新用户
  async updateUser(id: number, user: Partial<User>): Promise<User> {
    try {
      const response: AxiosResponse<User> = await apiClient.put(
        `/users/${id}`,
        user
      );
      return response.data;
    } catch (error) {
      throw new Error(`Failed to update user with id ${id}`);
    }
  },

  // 删除用户
  async deleteUser(id: number): Promise<void> {
    try {
      await apiClient.delete(`/users/${id}`);
    } catch (error) {
      throw new Error(`Failed to delete user with id ${id}`);
    }
  },

  // 搜索用户
  async searchUsers(query: string): Promise<User[]> {
    try {
      const response: AxiosResponse<User[]> = await apiClient.get(
        `/users?q=${query}`
      );
      return response.data;
    } catch (error) {
      throw new Error("Failed to search users");
    }
  },
};

// 通用 API 请求函数
export const fetchData = async <T>(url: string): Promise<T> => {
  try {
    const response: AxiosResponse<T> = await apiClient.get(url);
    return response.data;
  } catch (error) {
    throw new Error(`Failed to fetch data from ${url}`);
  }
};

// 带重试的请求函数
export const fetchWithRetry = async <T>(
  url: string,
  maxRetries: number = 3,
  delay: number = 1000
): Promise<T> => {
  let lastError: Error;

  for (let i = 0; i <= maxRetries; i++) {
    try {
      return await fetchData<T>(url);
    } catch (error) {
      lastError = error as Error;
      if (i < maxRetries) {
        await new Promise((resolve) => setTimeout(resolve, delay));
        delay *= 2; // 指数退避
      }
    }
  }

  throw lastError!;
};

// 模拟延迟的请求
export const fetchWithDelay = async <T>(
  data: T,
  delay: number = 1000
): Promise<T> => {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve(data);
    }, delay);
  });
};

// 模拟可能失败的请求
export const fetchWithRandomFailure = async <T>(
  data: T,
  failureRate: number = 0.3
): Promise<T> => {
  if (Math.random() < failureRate) {
    throw new Error("Random failure occurred");
  }
  return data;
};

export default apiClient;
