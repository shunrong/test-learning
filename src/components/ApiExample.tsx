/**
 * API 示例组件 - 用于演示异步测试和 Mock
 */

import React, { useState } from "react";
import { useFetch } from "@/hooks/useFetch";
import { userService } from "@/services/apiService";
import { User } from "@/types";

export const ApiExample: React.FC = () => {
  const [userId, setUserId] = useState<string>("1");
  const [users, setUsers] = useState<User[]>([]);
  const [selectedUser, setSelectedUser] = useState<User | null>(null);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);

  // 使用自定义 useFetch Hook
  const {
    data: fetchedUser,
    loading: fetchLoading,
    error: fetchError,
    execute: executeUserFetch,
  } = useFetch<User>(`https://jsonplaceholder.typicode.com/users/${userId}`, {
    immediate: false,
  });

  const handleFetchUser = async () => {
    await executeUserFetch();
  };

  const handleFetchUsers = async () => {
    setLoading(true);
    setError(null);
    try {
      const fetchedUsers = await userService.getUsers();
      setUsers(fetchedUsers);
    } catch (err) {
      setError(err instanceof Error ? err.message : "获取用户列表失败");
    } finally {
      setLoading(false);
    }
  };

  const handleGetUserById = async () => {
    if (!userId || isNaN(Number(userId))) {
      setError("请输入有效的用户 ID");
      return;
    }

    setLoading(true);
    setError(null);
    try {
      const user = await userService.getUserById(Number(userId));
      setSelectedUser(user);
    } catch (err) {
      setError(err instanceof Error ? err.message : "获取用户信息失败");
    } finally {
      setLoading(false);
    }
  };

  const handleCreateUser = async () => {
    setLoading(true);
    setError(null);
    try {
      const newUser = await userService.createUser({
        name: "新用户",
        email: "newuser@example.com",
      });
      setUsers((prev) => [...prev, newUser]);
    } catch (err) {
      setError(err instanceof Error ? err.message : "创建用户失败");
    } finally {
      setLoading(false);
    }
  };

  const containerStyle: React.CSSProperties = {
    border: "1px solid #ddd",
    borderRadius: "8px",
    padding: "16px",
    maxWidth: "600px",
  };

  const sectionStyle: React.CSSProperties = {
    marginBottom: "20px",
    padding: "12px",
    backgroundColor: "#f8f9fa",
    borderRadius: "4px",
  };

  const buttonStyle: React.CSSProperties = {
    padding: "8px 16px",
    margin: "4px",
    border: "none",
    borderRadius: "4px",
    cursor: "pointer",
    backgroundColor: "#74c0fc",
    color: "white",
  };

  const disabledButtonStyle: React.CSSProperties = {
    ...buttonStyle,
    backgroundColor: "#ddd",
    cursor: "not-allowed",
  };

  return (
    <div data-testid="api-example" style={containerStyle}>
      <h3>API 调用示例</h3>

      {/* 错误显示 */}
      {(error || fetchError) && (
        <div
          data-testid="error-message"
          style={{
            padding: "8px",
            backgroundColor: "#ffe6e6",
            color: "#d63031",
            borderRadius: "4px",
            marginBottom: "16px",
          }}
        >
          {error || fetchError?.message}
        </div>
      )}

      {/* 加载状态 */}
      {(loading || fetchLoading) && (
        <div
          data-testid="loading-indicator"
          style={{
            padding: "8px",
            backgroundColor: "#e3f2fd",
            color: "#1976d2",
            borderRadius: "4px",
            marginBottom: "16px",
          }}
        >
          加载中...
        </div>
      )}

      {/* 获取所有用户 */}
      <div style={sectionStyle}>
        <h4>获取所有用户</h4>
        <button
          data-testid="fetch-users-button"
          onClick={handleFetchUsers}
          disabled={loading}
          style={loading ? disabledButtonStyle : buttonStyle}
        >
          获取用户列表
        </button>
        {users.length > 0 && (
          <div data-testid="users-list" style={{ marginTop: "12px" }}>
            <p>获取到 {users.length} 个用户:</p>
            <ul style={{ maxHeight: "150px", overflowY: "auto" }}>
              {users.slice(0, 5).map((user) => (
                <li key={user.id} data-testid={`user-item-${user.id}`}>
                  {user.name} ({user.email})
                </li>
              ))}
              {users.length > 5 && <li>...</li>}
            </ul>
          </div>
        )}
      </div>

      {/* 根据 ID 获取用户 */}
      <div style={sectionStyle}>
        <h4>根据 ID 获取用户</h4>
        <div
          style={{
            display: "flex",
            alignItems: "center",
            gap: "8px",
            marginBottom: "8px",
          }}
        >
          <label>用户 ID:</label>
          <input
            data-testid="user-id-input"
            type="number"
            value={userId}
            onChange={(e) => setUserId(e.target.value)}
            style={{
              padding: "4px 8px",
              border: "1px solid #ddd",
              borderRadius: "4px",
              width: "100px",
            }}
          />
          <button
            data-testid="get-user-button"
            onClick={handleGetUserById}
            disabled={loading}
            style={loading ? disabledButtonStyle : buttonStyle}
          >
            获取用户
          </button>
        </div>
        {selectedUser && (
          <div data-testid="selected-user" style={{ marginTop: "12px" }}>
            <p>
              <strong>姓名:</strong> {selectedUser.name}
            </p>
            <p>
              <strong>邮箱:</strong> {selectedUser.email}
            </p>
          </div>
        )}
      </div>

      {/* 使用 useFetch Hook */}
      <div style={sectionStyle}>
        <h4>使用 useFetch Hook</h4>
        <button
          data-testid="use-fetch-button"
          onClick={handleFetchUser}
          disabled={fetchLoading}
          style={fetchLoading ? disabledButtonStyle : buttonStyle}
        >
          通过 Hook 获取用户
        </button>
        {fetchedUser && (
          <div data-testid="fetched-user" style={{ marginTop: "12px" }}>
            <p>
              <strong>姓名:</strong> {fetchedUser.name}
            </p>
            <p>
              <strong>邮箱:</strong> {fetchedUser.email}
            </p>
          </div>
        )}
      </div>

      {/* 创建用户 */}
      <div style={sectionStyle}>
        <h4>创建新用户</h4>
        <button
          data-testid="create-user-button"
          onClick={handleCreateUser}
          disabled={loading}
          style={loading ? disabledButtonStyle : buttonStyle}
        >
          创建用户
        </button>
      </div>
    </div>
  );
};
