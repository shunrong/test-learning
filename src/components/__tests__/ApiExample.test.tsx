/**
 * ApiExample 组件测试
 */

import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { ApiExample } from "../ApiExample";
import { userService } from "@/services/apiService";
import * as useFetchModule from "@/hooks/useFetch";

// Mock 依赖
jest.mock("@/services/apiService");
jest.mock("@/hooks/useFetch");

const mockUserService = userService as jest.Mocked<typeof userService>;
const mockUseFetch = useFetchModule.useFetch as jest.MockedFunction<
  typeof useFetchModule.useFetch
>;

// 模拟数据
const mockUser = {
  id: 1,
  name: "张三",
  email: "zhangsan@example.com",
};

const mockUsers = [
  { id: 1, name: "张三", email: "zhangsan@example.com" },
  { id: 2, name: "李四", email: "lisi@example.com" },
  { id: 3, name: "王五", email: "wangwu@example.com" },
];

describe("ApiExample组件", () => {
  const mockExecute = jest.fn();
  const mockReset = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();

    // 默认 useFetch mock
    mockUseFetch.mockReturnValue({
      data: null,
      loading: false,
      error: null,
      execute: mockExecute,
      reset: mockReset,
    });
  });

  describe("基础渲染", () => {
    it("应该正确渲染 API 示例组件", () => {
      render(<ApiExample />);

      expect(screen.getByTestId("api-example")).toBeInTheDocument();
      expect(screen.getByText("API 调用示例")).toBeInTheDocument();
      expect(screen.getByTestId("fetch-users-button")).toBeInTheDocument();
      expect(screen.getByTestId("get-user-button")).toBeInTheDocument();
      expect(screen.getByTestId("use-fetch-button")).toBeInTheDocument();
      expect(screen.getByTestId("create-user-button")).toBeInTheDocument();
    });

    it("应该显示用户 ID 输入框和默认值", () => {
      render(<ApiExample />);

      const userIdInput = screen.getByTestId("user-id-input");
      expect(userIdInput).toBeInTheDocument();
      expect(userIdInput).toHaveValue(1);
    });

    it("应该正确渲染各个功能区域", () => {
      render(<ApiExample />);

      expect(screen.getByText("获取所有用户")).toBeInTheDocument();
      expect(screen.getByText("根据 ID 获取用户")).toBeInTheDocument();
      expect(screen.getByText("使用 useFetch Hook")).toBeInTheDocument();
      expect(screen.getByText("创建新用户")).toBeInTheDocument();
    });
  });

  describe("获取用户列表功能", () => {
    it("应该能够成功获取用户列表", async () => {
      const user = userEvent.setup();
      mockUserService.getUsers.mockResolvedValue(mockUsers);

      render(<ApiExample />);

      const fetchUsersButton = screen.getByTestId("fetch-users-button");
      await user.click(fetchUsersButton);

      await waitFor(() => {
        expect(mockUserService.getUsers).toHaveBeenCalledTimes(1);
      });

      await waitFor(() => {
        expect(screen.getByTestId("users-list")).toBeInTheDocument();
        expect(screen.getByText("获取到 3 个用户:")).toBeInTheDocument();
        expect(
          screen.getByText("张三 (zhangsan@example.com)")
        ).toBeInTheDocument();
        expect(screen.getByText("李四 (lisi@example.com)")).toBeInTheDocument();
      });
    });

    it("应该在获取用户列表时显示加载状态", async () => {
      const user = userEvent.setup();
      mockUserService.getUsers.mockImplementation(
        () =>
          new Promise((resolve) => setTimeout(() => resolve(mockUsers), 100))
      );

      render(<ApiExample />);

      const fetchUsersButton = screen.getByTestId("fetch-users-button");
      await user.click(fetchUsersButton);

      expect(screen.getByTestId("loading-indicator")).toBeInTheDocument();
      expect(screen.getByText("加载中...")).toBeInTheDocument();

      await waitFor(() => {
        expect(
          screen.queryByTestId("loading-indicator")
        ).not.toBeInTheDocument();
      });
    });

    it("应该处理获取用户列表的错误", async () => {
      const user = userEvent.setup();
      const errorMessage = "网络错误";
      mockUserService.getUsers.mockRejectedValue(new Error(errorMessage));

      render(<ApiExample />);

      const fetchUsersButton = screen.getByTestId("fetch-users-button");
      await user.click(fetchUsersButton);

      await waitFor(() => {
        expect(screen.getByTestId("error-message")).toBeInTheDocument();
        expect(screen.getByText(errorMessage)).toBeInTheDocument();
      });
    });

    it("应该限制显示的用户数量不超过5个", async () => {
      const user = userEvent.setup();
      const manyUsers = Array.from({ length: 10 }, (_, i) => ({
        id: i + 1,
        name: `用户${i + 1}`,
        email: `user${i + 1}@example.com`,
      }));

      mockUserService.getUsers.mockResolvedValue(manyUsers);

      render(<ApiExample />);

      const fetchUsersButton = screen.getByTestId("fetch-users-button");
      await user.click(fetchUsersButton);

      await waitFor(() => {
        expect(screen.getByText("获取到 10 个用户:")).toBeInTheDocument();

        // 应该只显示前5个用户
        expect(
          screen.getByText("用户1 (user1@example.com)")
        ).toBeInTheDocument();
        expect(
          screen.getByText("用户5 (user5@example.com)")
        ).toBeInTheDocument();

        // 应该显示省略号
        expect(screen.getByText("...")).toBeInTheDocument();
      });
    });
  });

  describe("根据ID获取用户功能", () => {
    it("应该能够根据ID成功获取用户", async () => {
      const user = userEvent.setup();
      mockUserService.getUserById.mockResolvedValue(mockUser);

      render(<ApiExample />);

      const userIdInput = screen.getByTestId("user-id-input");
      const getUserButton = screen.getByTestId("get-user-button");

      await user.clear(userIdInput);
      await user.type(userIdInput, "1");
      await user.click(getUserButton);

      await waitFor(() => {
        expect(mockUserService.getUserById).toHaveBeenCalledWith(1);
      });

      await waitFor(() => {
        expect(screen.getByTestId("selected-user")).toBeInTheDocument();
        expect(screen.getByText("张三")).toBeInTheDocument();
        expect(screen.getByText("zhangsan@example.com")).toBeInTheDocument();
      });
    });

    it("应该验证用户ID的有效性", async () => {
      const user = userEvent.setup();

      render(<ApiExample />);

      const userIdInput = screen.getByTestId("user-id-input");
      const getUserButton = screen.getByTestId("get-user-button");

      await user.clear(userIdInput);
      await user.type(userIdInput, "abc");
      await user.click(getUserButton);

      await waitFor(() => {
        expect(screen.getByTestId("error-message")).toBeInTheDocument();
        expect(screen.getByText("请输入有效的用户 ID")).toBeInTheDocument();
      });

      expect(mockUserService.getUserById).not.toHaveBeenCalled();
    });

    it("应该处理空的用户ID", async () => {
      const user = userEvent.setup();

      render(<ApiExample />);

      const userIdInput = screen.getByTestId("user-id-input");
      const getUserButton = screen.getByTestId("get-user-button");

      await user.clear(userIdInput);
      await user.click(getUserButton);

      await waitFor(() => {
        expect(screen.getByTestId("error-message")).toBeInTheDocument();
        expect(screen.getByText("请输入有效的用户 ID")).toBeInTheDocument();
      });

      expect(mockUserService.getUserById).not.toHaveBeenCalled();
    });

    it("应该处理获取用户的错误", async () => {
      const user = userEvent.setup();
      const errorMessage = "用户不存在";
      mockUserService.getUserById.mockRejectedValue(new Error(errorMessage));

      render(<ApiExample />);

      const getUserButton = screen.getByTestId("get-user-button");
      await user.click(getUserButton);

      await waitFor(() => {
        expect(screen.getByTestId("error-message")).toBeInTheDocument();
        expect(screen.getByText(errorMessage)).toBeInTheDocument();
      });
    });
  });

  describe("useFetch Hook 功能", () => {
    it("应该使用 useFetch Hook 获取用户", async () => {
      const user = userEvent.setup();
      mockUseFetch.mockReturnValue({
        data: mockUser,
        loading: false,
        error: null,
        execute: mockExecute,
        reset: mockReset,
      });

      render(<ApiExample />);

      const useFetchButton = screen.getByTestId("use-fetch-button");
      await user.click(useFetchButton);

      expect(mockExecute).toHaveBeenCalledTimes(1);
      expect(screen.getByTestId("fetched-user")).toBeInTheDocument();
      expect(screen.getByText("张三")).toBeInTheDocument();
    });

    it("应该显示 useFetch 的加载状态", () => {
      mockUseFetch.mockReturnValue({
        data: null,
        loading: true,
        error: null,
        execute: mockExecute,
        reset: mockReset,
      });

      render(<ApiExample />);

      expect(screen.getByTestId("loading-indicator")).toBeInTheDocument();

      const useFetchButton = screen.getByTestId("use-fetch-button");
      expect(useFetchButton).toBeDisabled();
    });

    it("应该显示 useFetch 的错误状态", () => {
      const error = new Error("Hook 错误");
      mockUseFetch.mockReturnValue({
        data: null,
        loading: false,
        error,
        execute: mockExecute,
        reset: mockReset,
      });

      render(<ApiExample />);

      expect(screen.getByTestId("error-message")).toBeInTheDocument();
      expect(screen.getByText("Hook 错误")).toBeInTheDocument();
    });
  });

  describe("创建用户功能", () => {
    it("应该能够创建新用户", async () => {
      const user = userEvent.setup();
      const newUser = {
        id: 4,
        name: "新用户",
        email: "newuser@example.com",
      };

      mockUserService.createUser.mockResolvedValue(newUser);
      mockUserService.getUsers.mockResolvedValue(mockUsers);

      render(<ApiExample />);

      // 先获取用户列表
      const fetchUsersButton = screen.getByTestId("fetch-users-button");
      await user.click(fetchUsersButton);

      await waitFor(() => {
        expect(screen.getByText("获取到 3 个用户:")).toBeInTheDocument();
      });

      // 创建新用户
      const createUserButton = screen.getByTestId("create-user-button");
      await user.click(createUserButton);

      await waitFor(() => {
        expect(mockUserService.createUser).toHaveBeenCalledWith({
          name: "新用户",
          email: "newuser@example.com",
        });
      });
    });

    it("应该处理创建用户的错误", async () => {
      const user = userEvent.setup();
      const errorMessage = "创建失败";
      mockUserService.createUser.mockRejectedValue(new Error(errorMessage));

      render(<ApiExample />);

      const createUserButton = screen.getByTestId("create-user-button");
      await user.click(createUserButton);

      await waitFor(() => {
        expect(screen.getByTestId("error-message")).toBeInTheDocument();
        expect(screen.getByText(errorMessage)).toBeInTheDocument();
      });
    });
  });

  describe("加载状态管理", () => {
    it("在加载时应该禁用所有按钮", async () => {
      const user = userEvent.setup();
      mockUserService.getUsers.mockImplementation(
        () =>
          new Promise((resolve) => setTimeout(() => resolve(mockUsers), 100))
      );

      render(<ApiExample />);

      const fetchUsersButton = screen.getByTestId("fetch-users-button");
      await user.click(fetchUsersButton);

      // 检查所有按钮都被禁用
      expect(screen.getByTestId("fetch-users-button")).toBeDisabled();
      expect(screen.getByTestId("get-user-button")).toBeDisabled();
      expect(screen.getByTestId("create-user-button")).toBeDisabled();

      await waitFor(() => {
        expect(
          screen.queryByTestId("loading-indicator")
        ).not.toBeInTheDocument();
      });
    });
  });

  describe("错误处理", () => {
    it("应该清除之前的错误信息", async () => {
      const user = userEvent.setup();

      // 先触发一个错误
      mockUserService.getUsers.mockRejectedValue(new Error("第一个错误"));

      render(<ApiExample />);

      const fetchUsersButton = screen.getByTestId("fetch-users-button");
      await user.click(fetchUsersButton);

      await waitFor(() => {
        expect(screen.getByText("第一个错误")).toBeInTheDocument();
      });

      // 然后成功请求
      mockUserService.getUsers.mockResolvedValue(mockUsers);
      await user.click(fetchUsersButton);

      await waitFor(() => {
        expect(screen.queryByText("第一个错误")).not.toBeInTheDocument();
        expect(screen.getByTestId("users-list")).toBeInTheDocument();
      });
    });

    it("应该处理非 Error 对象的错误", async () => {
      const user = userEvent.setup();
      mockUserService.getUsers.mockRejectedValue("字符串错误");

      render(<ApiExample />);

      const fetchUsersButton = screen.getByTestId("fetch-users-button");
      await user.click(fetchUsersButton);

      await waitFor(() => {
        expect(screen.getByText("获取用户列表失败")).toBeInTheDocument();
      });
    });
  });

  describe("可访问性", () => {
    it("所有交互元素应该有正确的 data-testid", () => {
      render(<ApiExample />);

      expect(screen.getByTestId("api-example")).toBeInTheDocument();
      expect(screen.getByTestId("fetch-users-button")).toBeInTheDocument();
      expect(screen.getByTestId("user-id-input")).toBeInTheDocument();
      expect(screen.getByTestId("get-user-button")).toBeInTheDocument();
      expect(screen.getByTestId("use-fetch-button")).toBeInTheDocument();
      expect(screen.getByTestId("create-user-button")).toBeInTheDocument();
    });

    it("用户列表项应该有正确的 data-testid", async () => {
      const user = userEvent.setup();
      mockUserService.getUsers.mockResolvedValue(mockUsers);

      render(<ApiExample />);

      const fetchUsersButton = screen.getByTestId("fetch-users-button");
      await user.click(fetchUsersButton);

      await waitFor(() => {
        mockUsers.forEach((userData) => {
          expect(
            screen.getByTestId(`user-item-${userData.id}`)
          ).toBeInTheDocument();
        });
      });
    });
  });

  describe("快照测试", () => {
    it("应该匹配初始状态的快照", () => {
      const { container } = render(<ApiExample />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it("应该匹配有数据状态的快照", async () => {
      const user = userEvent.setup();
      mockUserService.getUsers.mockResolvedValue(mockUsers);

      const { container } = render(<ApiExample />);

      const fetchUsersButton = screen.getByTestId("fetch-users-button");
      await user.click(fetchUsersButton);

      await waitFor(() => {
        expect(screen.getByTestId("users-list")).toBeInTheDocument();
      });

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe("输入值变化", () => {
    it("应该正确更新用户ID输入值", async () => {
      const user = userEvent.setup();

      render(<ApiExample />);

      const userIdInput = screen.getByTestId("user-id-input");

      await user.clear(userIdInput);
      await user.type(userIdInput, "123");

      expect(userIdInput).toHaveValue(123);
    });
  });
});
