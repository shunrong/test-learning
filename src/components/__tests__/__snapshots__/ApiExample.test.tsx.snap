// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiExample组件 快照测试 应该匹配初始状态的快照 1`] = `
<div
  data-testid="api-example"
  style="border: 1px solid #ddd; border-radius: 8px; padding: 16px; max-width: 600px;"
>
  <h3>
    API 调用示例
  </h3>
  <div
    style="margin-bottom: 20px; padding: 12px; background-color: rgb(248, 249, 250); border-radius: 4px;"
  >
    <h4>
      获取所有用户
    </h4>
    <button
      data-testid="fetch-users-button"
      style="padding: 8px 16px; margin: 4px; border-radius: 4px; cursor: pointer; background-color: rgb(116, 192, 252); color: white;"
    >
      获取用户列表
    </button>
  </div>
  <div
    style="margin-bottom: 20px; padding: 12px; background-color: rgb(248, 249, 250); border-radius: 4px;"
  >
    <h4>
      根据 ID 获取用户
    </h4>
    <div
      style="display: flex; align-items: center; gap: 8px; margin-bottom: 8px;"
    >
      <label>
        用户 ID:
      </label>
      <input
        data-testid="user-id-input"
        style="padding: 4px 8px; border: 1px solid #ddd; border-radius: 4px; width: 100px;"
        type="number"
        value="1"
      />
      <button
        data-testid="get-user-button"
        style="padding: 8px 16px; margin: 4px; border-radius: 4px; cursor: pointer; background-color: rgb(116, 192, 252); color: white;"
      >
        获取用户
      </button>
    </div>
  </div>
  <div
    style="margin-bottom: 20px; padding: 12px; background-color: rgb(248, 249, 250); border-radius: 4px;"
  >
    <h4>
      使用 useFetch Hook
    </h4>
    <button
      data-testid="use-fetch-button"
      style="padding: 8px 16px; margin: 4px; border-radius: 4px; cursor: pointer; background-color: rgb(116, 192, 252); color: white;"
    >
      通过 Hook 获取用户
    </button>
  </div>
  <div
    style="margin-bottom: 20px; padding: 12px; background-color: rgb(248, 249, 250); border-radius: 4px;"
  >
    <h4>
      创建新用户
    </h4>
    <button
      data-testid="create-user-button"
      style="padding: 8px 16px; margin: 4px; border-radius: 4px; cursor: pointer; background-color: rgb(116, 192, 252); color: white;"
    >
      创建用户
    </button>
  </div>
</div>
`;

exports[`ApiExample组件 快照测试 应该匹配有数据状态的快照 1`] = `
<div
  data-testid="api-example"
  style="border: 1px solid #ddd; border-radius: 8px; padding: 16px; max-width: 600px;"
>
  <h3>
    API 调用示例
  </h3>
  <div
    style="margin-bottom: 20px; padding: 12px; background-color: rgb(248, 249, 250); border-radius: 4px;"
  >
    <h4>
      获取所有用户
    </h4>
    <button
      data-testid="fetch-users-button"
      style="padding: 8px 16px; margin: 4px; border-radius: 4px; cursor: pointer; background-color: rgb(116, 192, 252); color: white;"
    >
      获取用户列表
    </button>
    <div
      data-testid="users-list"
      style="margin-top: 12px;"
    >
      <p>
        获取到 
        3
         个用户:
      </p>
      <ul
        style="max-height: 150px; overflow-y: auto;"
      >
        <li
          data-testid="user-item-1"
        >
          张三
           (
          zhangsan@example.com
          )
        </li>
        <li
          data-testid="user-item-2"
        >
          李四
           (
          lisi@example.com
          )
        </li>
        <li
          data-testid="user-item-3"
        >
          王五
           (
          wangwu@example.com
          )
        </li>
      </ul>
    </div>
  </div>
  <div
    style="margin-bottom: 20px; padding: 12px; background-color: rgb(248, 249, 250); border-radius: 4px;"
  >
    <h4>
      根据 ID 获取用户
    </h4>
    <div
      style="display: flex; align-items: center; gap: 8px; margin-bottom: 8px;"
    >
      <label>
        用户 ID:
      </label>
      <input
        data-testid="user-id-input"
        style="padding: 4px 8px; border: 1px solid #ddd; border-radius: 4px; width: 100px;"
        type="number"
        value="1"
      />
      <button
        data-testid="get-user-button"
        style="padding: 8px 16px; margin: 4px; border-radius: 4px; cursor: pointer; background-color: rgb(116, 192, 252); color: white;"
      >
        获取用户
      </button>
    </div>
  </div>
  <div
    style="margin-bottom: 20px; padding: 12px; background-color: rgb(248, 249, 250); border-radius: 4px;"
  >
    <h4>
      使用 useFetch Hook
    </h4>
    <button
      data-testid="use-fetch-button"
      style="padding: 8px 16px; margin: 4px; border-radius: 4px; cursor: pointer; background-color: rgb(116, 192, 252); color: white;"
    >
      通过 Hook 获取用户
    </button>
  </div>
  <div
    style="margin-bottom: 20px; padding: 12px; background-color: rgb(248, 249, 250); border-radius: 4px;"
  >
    <h4>
      创建新用户
    </h4>
    <button
      data-testid="create-user-button"
      style="padding: 8px 16px; margin: 4px; border-radius: 4px; cursor: pointer; background-color: rgb(116, 192, 252); color: white;"
    >
      创建用户
    </button>
  </div>
</div>
`;
